src/vanilla/utils/atomWithObservable.ts

```diff
@@ -81,107 +74,86 @@ export function atomWithObservable<Data>(
       observable = itself
     }
 
-    let resolve: ((result: Result) => void) | undefined
-    const makePending = () =>
-      new Promise<Result>((r) => {
-        resolve = r
-      })
-    const initialResult: Result | Promise<Result> =
+    // TODO: replace with https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers
+    // once CI Node version is bump to 22+
+    let resolve: (value: Data | PromiseLike<Data>) => void
+    let reject: (reason?: any) => void
+    const promise = new Promise<Data>((_resolve, _reject) => {
+      resolve = _resolve
+      reject = _reject
+    })
+
+    const initialResult: Data | Promise<Data> =
       options && 'initialValue' in options
-        ? {
-            d:
-              typeof options.initialValue === 'function'
-                ? (options.initialValue as () => Data)()
-                : (options.initialValue as Data),
-          }
-        : makePending()
-
-    let setResult: ((result: Result) => void) | undefined
-    let lastResult: Result | undefined
-    const listener = (result: Result) => {
-      lastResult = result
-      resolve?.(result)
-      setResult?.(result)
-    }
+        ? typeof options.initialValue === 'function'
+          ? (options.initialValue as () => Data)()
+          : (options.initialValue as Data)
+        : promise
+
+    const resultAtom = atom(initialResult)
+
+    let preMountData: Data | typeof EMPTY = EMPTY
+
+    let setResult: ((result: Data | Promise<Data>) => void) | undefined
 
     let subscription: Subscription | undefined
-    let timer: Timeout | undefined
-    const isNotMounted = () => !setResult
-    const start = () => {
-      if (subscription) {
-        clearTimeout(timer)
-        subscription.unsubscribe()
-      }
+    const startSubscription = () => {
+      subscription?.unsubscribe()
       subscription = observable.subscribe({
-        next: (d) => listener({ d }),
-        error: (e) => listener({ e }),
+        next: (value) => {
+          resolve(value)
+          preMountData = value
+          setResult?.(value)
+        },
+        error: (error) => {
+          reject(error)
+          if (error instanceof Error) {
+            // @ts-expect-error TypeScript doesn't allow setting atom to error
+            // If we use Promise.reject here instead, Vitest will complain about an uncaught Error
+            setResult?.(error)
```

Comment:
We can't distinguish error and value. Theoretically, value can be `Error`.
Please consider to revert to `{ d: Data } | { e: AnyError }`

Reply:
For that, I did the below part: `setResult?.(Promise.reject(error))`. Maybe we should always do that?

Reply:
Okay, I've cleaned up this bit and also added a test for it 🙏